os/drivers/input/ist415.c

```diff
@@ -0,0 +1,407 @@
+	size_t outlen;
+	irqstate_t flags;
+	int ret;
+
+	DEBUGASSERT(buffer);
+	DEBUGASSERT(buflen > 0);
+	DEBUGASSERT(filep);
+	inode = filep->f_inode;
+
+	DEBUGASSERT(inode && inode->i_private);
+	priv = inode->i_private;
+
+	/* Wait for semaphore to prevent concurrent reads */
+	touch_semtake(&priv->sem, false);
+
+	/* Read the touch data, only if screen has been touched or if we're waiting for touch up */
+	outlen = sizeof(struct touch_sample_s);
+	if (priv->int_pending) {
+		flags = enter_critical_section();
+		priv->int_pending = false;
+		leave_critical_section(flags);
+		ret = ist415_get_touch_data(priv, buffer);
+	}
+	touch_semgive(&priv->sem);
+	return ret < 0 ? ret : outlen;
+}
+
+static ssize_t ist415_write(FAR struct file *filep, FAR char *buffer, size_t buflen)
+{
+	//TODO
+	return 0;
+}
+
+#ifndef CONFIG_DISABLE_POLL
+static void ist415_pollnotify(FAR struct ist415_dev_s *dev, pollevent_t eventset)
+{
+	int itr;
+	for (itr = 0; itr < CONFIG_SERIAL_NPOLLWAITERS; itr++) {
+		struct pollfd *fds = dev->fds[itr];
+		if (fds) {
+			fds->revents |= (fds->events & eventset);
+			if (fds->revents != 0) {
+				sem_post(fds->sem);
+			}
+		}
+	}
+}
+
+static int ist415_poll(FAR struct file *filep, FAR struct pollfd *fds, bool setup)
+{
+	FAR struct inode *inode = filep->f_inode;
+	FAR struct ist415_dev_s *dev = inode->i_private;
+	pollevent_t eventset;
+	int ret;
+	int i;
+	bool pending = false;
+	/* Some sanity checking */
+#ifdef CONFIG_DEBUG
+	if (!dev || !fds) {
+		return -ENODEV;
+	}
+#endif
+	ret = touch_semtake(&dev->pollsem, true);
+	if (ret < 0) {
+		/* A signal received while waiting for access to the poll data
+		 * will abort the operation.
+		 */
+		return ret;
+	}
+
+	if (setup) {
+		/* This is a request to set up the poll. Find an available
+		 * slot for the poll structure reference
+		 */
+		for (i = 0; i < CONFIG_SERIAL_NPOLLWAITERS; i++) {
+			/* Find an available slot */
+			if (!dev->fds[i]) {
+				/* Bind the poll structure and this slot */
+				dev->fds[i] = fds;
+				fds->priv = &dev->fds[i];
+				fds->filep = (void *)filep;
+				break;
+			}
+		}
+
+		if (i >= CONFIG_SERIAL_NPOLLWAITERS) {
+			fds->priv = NULL;
+			fds->filep = NULL;
+			ret = -EBUSY;
+			goto errout;
+		}
+
+		eventset = 0;
+		touch_semtake(&dev->sem, false);
+		pending = dev->int_pending;
+
+		if (pending) {
+			eventset |= (fds->events & POLLIN);
+		}
+		touch_semgive(&dev->sem);
+		if (eventset) {
+			ist415_pollnotify(dev, POLLIN);
+		}
+
+	} else if (fds->priv) {
+		/* This is a request to tear down the poll. */
+		struct pollfd **slot = (struct pollfd **)fds->priv;
+		/* Remove all memory of the poll setup */
+		*slot = NULL;
+		fds->priv = NULL;
+		fds->filep = NULL;
+	}
+errout:
+	touch_semgive(&dev->pollsem);
+	return ret;
+}
+#endif
+
+static void touch_interrupt(int d)
+{
+	FAR struct ist415_dev_s *priv = &g_ist415_priv;
+	irqstate_t state = enter_critical_section();
+	priv->int_pending = true;
+	leave_critical_section(state);
```

Comment:
hmmm. only for line 368, do we need to call enter / leave critical section? 
I think without them, it will be the same.